src/backend/distributed/executor/local_executor.c

```diff
@@ -405,10 +406,11 @@ ShouldExecuteTasksLocally(List *taskList)
 		 * Still, we'll be avoding the network round trip for this node.
 		 *
 		 * Note that we shouldn't use local execution if any distributed execution
-		 * has happened because that'd break transaction visibility rules and
-		 * many other things.
+		 * modified a distributed table because that'd break transaction visibility
+		 * rules and many other things. Therefore it is the callers responsibility
+		 * to make sure no modifications are made in a distributed transaction
 		 */
-		return !AnyConnectionAccessedPlacements();
+		return !TransactionModifiedDistributedTable(execution);
```

Comment:
Would it be sufficient to just check `XactModificationLevel == XACT_MODIFICATION_DATA`?

Reply:
I think no because that flag is also set after local execution modified any placement. The basic idea of that check is to make sure that no remote access happened before, excluding local accesses

Reply:
Stepping back. I think the main thing we wanted to achieve here is to not use local execution if we already connected to localhost (e.g. to do DDL). Should we add a `ConnectedToLocalhost` flag that becomes true if we're opening a connection to groupId == GetLocalGroupId()? Eventually this flag will disappear once we support local execution for everything.